CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

IntegralProgrammer opened the pull request:
This Pull Request moves the `SlackNotificationsTask` and `NightlySlackNotifications` classes out of the `patient-portal` module and into their own new `slack-notifications` module thus allowing for the configuration of periodic performance metric alerting via Slack on projects that do not use the `patient-portal` module but still generate performance metrics (ie. _HERACLES_).

Testing Instructions
--------------------------

1. Build this (`CARDS-2272`) branch, including a Docker image, with `mvn clean install -Pdocker`.
2. In a new terminal window, enter the `Utilities/Development` directory and run `nodejs mock_slack.js`
3. In the `compose-cluster` directory, run:
    - `export SLACK_PERFORMANCE_URL=http://172.99.0.1:8000`
    - `export NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE='0 * * ? * *'`
    - `python3 generate_compose_yaml.py --dev_docker_image --oak_filesystem --composum --cards_project cards4prems --slack_notifications --subnet 172.99.0.0/16`
    - `docker-compose build && docker-compose up -d`
4. Monitor the terminal running `mock_slack.js` - performance update messages should be sent every minute
5. Stop the Docker Compose environment and clean up with `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
6. In the root of the `cards` directory, run:
    - `export SLACK_PERFORMANCE_URL=http://localhost:8000`
    - `export NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE='0 * * ? * *'`
    - `./start_cards.sh --dev --project cards4prems --slack_notifications`
7. Monitor the terminal running `mock_slack.js` - performance update messages should be sent every minute

IntegralProgrammer committed:
CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

Slack notifications will be enabled whenever the SLACK_NOTIFICATIONS_ENABLED=true environment variable is passed to the CARDS Docker container

IntegralProgrammer committed:
CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

Slack notifications will be enabled whenever the --slack_notifications flag is passed to start_cards.sh

IntegralProgrammer committed:
CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

Remove NightlySlackNotifications and SlackNotificationsTask from the patient-portal module

IntegralProgrammer committed:
CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

Remove NightlySlackNotifications and SlackNotificationsTask from the patient-portal module

IntegralProgrammer committed:
CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

Slack notifications will be enabled whenever the --slack_notifications flag is passed to generate_compose_yaml.py

sdumitriu commented on modules/slack-notifications/src/main/java/io/uhndata/cards/slacknotifications/NightlySlackNotifications.java:
I'd like to phase-out using `nightly` in class names, since we could schedule these to run weekly or every hour or during daytime. How about `SlackNotificationsScheduler`?

IntegralProgrammer replied:
I renamed the class to `ScheduledSlackNotifications` in https://github.com/data-team-uhn/cards/pull/1555/commits/36226d4afe214163a59b8e67ac64d97a1dbf03f5 :heavy_check_mark:.

sdumitriu commented on modules/slack-notifications/src/main/features/feature.json:
```diff
@@ -0,0 +1,40 @@
+  "configurations":{
+    "org.apache.sling.serviceusermapping.impl.ServiceUserMapperImpl.amended~slackNotifications":{
+      "user.mapping":[
+        "io.uhndata.cards.slack-notifications:SlackNotifications=[cards-metrics]"
```
There's an equivalent line in the patient portal feature json which needs to be removed.

IntegralProgrammer replied:
Fixed in https://github.com/data-team-uhn/cards/pull/1555/commits/076c1519c21b988d5c90ea3042fd31da56ea5ef6 :heavy_check_mark:.

IntegralProgrammer committed:
CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

Rename NightlySlackNotifications to ScheduledSlackNotifications as the notifications do not necessarily need to run nightly

IntegralProgrammer committed:
CARDS-2272: SlackNotificationsTask and NightlySlackNotifications should be moved into their own CARDS module

Remove the io.uhndata.cards.patient-portal:SlackNotifications=[cards-metrics] org.apache.sling.serviceusermapping entry in the patient-portal feature.json file
